### PR TITLE
feature(cy): add BDQ algorithm

### PR DESCRIPTION
**Pull request by Cloud-Pku**
## Description
Add the bdq algorithm and config for hopper and halfcheetah env.

## TODO
n-step for BDQ
gradient scaling for backbone network
unify bdq_nstep_td_error and q_nstep_td_error function
unify action_map function
unit Tests writing

## Check List

- [x] merge the latest version source branch/repo, and resolve all the conflicts
- [x] pass style check
- [x] pass all the tests

### CODE REVIEW

File: ding/model/common/head.py [hunk 1]
```diff
@@ -174,6 +174,108 @@ def forward(self, x: torch.Tensor) -> Dict:
...
+            activation: Optional[nn.Module] = nn.ReLU(),
+            noise: Optional[bool] = False,
+    ) -> None:
+        """
+        Overview:
+            Init the ``BranchingHead`` layers according to the provided arguments.
+        Arguments:
+            - hidden_size (:obj:`int`): The ``hidden_size`` of the MLP connected to ``BranchingHead``.
+            - num_branches (:obj:`int`): The number of branches, which is equivalent to the action dimension.
+            - action_per_branch (:obj:`int`): The number of actions in each dimension.
```
Comment (puyuan1996) [inline ding/model/common/head.py:197]: - action_bins_per_branch (:obj:`int`): The number of action bins in each dimension.

action_per_branch -> action_bins_per_branch 更加清晰一些？

File: ding/policy/bdq.py [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+
+from .base_policy import Policy
+from .common_utils import default_preprocess_learn
+
+
+@POLICY_REGISTRY.register('bdq')
+class BDQPolicy(Policy):
+    r"""
+    Overview:
+        Policy class of BDQ algorithm, extended by PER/multi-step TD.
```
Comment (puyuan1996) [inline ding/policy/bdq.py:20]: add paper links and key idea descriptions?

File: ding/policy/bdq.py [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+            'total_loss': loss.item(),
+            'q_value': q_value.mean().item(),
+            'target_q_value': target_q_value.mean().item(),
+            'priority': td_error_per_sample.abs().tolist(),
+            # Only discrete action satisfying len(data['action'])==1 can return this and draw histogram on tensorboard.
+            # '[histogram]action_distribution': data['action'],
+        }
+
+    def _monitor_vars_learn(self) -> List[str]:
+        return ['cur_lr', 'total_loss', 'q_value']
```
Comment (puyuan1996) [inline ding/policy/bdq.py:242]: add more tb statistics?

File: ding/rl_utils/td.py [hunk 4]
```diff
@@ -453,6 +453,62 @@ def q_nstep_td_error(
...
+        data: namedtuple,
+        gamma: Union[float, list],
+        nstep: int = 1,
+        cum_reward: bool = False,
+        value_gamma: Optional[torch.Tensor] = None,
+        criterion: torch.nn.modules = nn.MSELoss(reduction='none'),
+) -> torch.Tensor:
+    """
+    Overview:
+        Multistep (1 step or n step) td_error for BDQ algorithm
```
Comment (puyuan1996) [inline ding/rl_utils/td.py:466]: add paper links and td equation descriptions?

File: dizoo/mujoco/config/halfcheetah_bdq_config.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+halfcheetah_bdq_config = EasyDict(halfcheetah_bdq_config)
+main_config = halfcheetah_bdq_config
+
+halfcheetah_bdq_create_config = dict(
+    env=dict(
+        type='mujoco',
+        import_names=['dizoo.mujoco.envs.mujoco_env'],
+    ),
+    env_manager=dict(type='subprocess'),
+    # env_manager=dict(type='subprocess'),
```
Comment (puyuan1996) [inline dizoo/mujoco/config/halfcheetah_bdq_config.py:62]: delete unsed commets

File: dizoo/mujoco/envs/mujoco_env.py [hunk 6]
```diff
@@ -39,6 +39,10 @@ def __init__(self, cfg: dict) -> None:
         self._replay_path = None
         self._replay_path_gif = cfg.replay_path_gif
         self._save_replay_gif = cfg.save_replay_gif
+        self._action_per_branch = cfg.action_per_branch if 'action_per_branch' in cfg else None
```
Comment (puyuan1996) [inline dizoo/mujoco/envs/mujoco_env.py:42]: could use default_config:

    config = dict(
        action_clip=False,
        delay_reward_step=0,
        replay_path=None,
        save_replay_gif=False,
        replay_path_gif=None,
        action_per_branch=None,
    )

File: dizoo/mujoco/envs/mujoco_env.py [hunk 7]
```diff
@@ -39,6 +39,10 @@ def __init__(self, cfg: dict) -> None:
...
+
+    def map_action(self, action: Union[np.ndarray, list]) -> Union[np.ndarray, list]:
+        return [2 * x / (self._action_per_branch - 1) - 1 for x in action]
```
Comment (puyuan1996) [inline dizoo/mujoco/envs/mujoco_env.py:45]: add annotations and examples

File: ding/model/common/head.py [hunk 8]
```diff
@@ -174,6 +174,108 @@ def forward(self, x: torch.Tensor) -> Dict:
...
+            layer_num: int = 1,
+            a_layer_num: Optional[int] = None,
+            v_layer_num: Optional[int] = None,
+            norm_type: Optional[str] = None,
+            activation: Optional[nn.Module] = nn.ReLU(),
+            noise: Optional[bool] = False,
+    ) -> None:
+        """
+        Overview:
+            Init the ``BranchingHead`` layers according to the provided arguments.
```
Comment (PaParaZz1) [inline ding/model/common/head.py:193]: add more explanation about the scenes of using `branch`

File: ding/model/common/head.py [hunk 9]
```diff
@@ -174,6 +174,108 @@ def forward(self, x: torch.Tensor) -> Dict:
...
+        value_out = torch.unsqueeze(value_out, 1)
+        action_out = []
+        for b in self.branches:
+            action_out.append(b(x))
+        action_scores = torch.stack(action_out, 1)
+        action_scores = action_scores - torch.mean(action_scores, 2, keepdim=True)
+        logits = value_out + action_scores
+        return {'logit': logits}
+
+
```
Comment (PaParaZz1) [inline ding/model/common/head.py:278]: add unittest for this file

File: ding/model/common/head.py [hunk 10]
```diff
@@ -174,6 +174,108 @@ def forward(self, x: torch.Tensor) -> Dict:
...
+            - outputs (:obj:`Dict`): Dict containing keyword ``logit`` (:obj:`torch.Tensor`).
+        Shapes:
+            - x: :math:`(B, N)`, where ``B = batch_size`` and ``N = hidden_size``.
+            - logit: :math:`(B, M)`, where ``M = output_size``.
+
+        Examples:
+            >>> head = BranchingHead(64, 5, 2)
+            >>> inputs = torch.randn(4, 64)
+            >>> outputs = head(inputs)
+            >>> assert isinstance(outputs, dict) and outputs['logit'].shape == torch.Size([4, 10])
```
Comment (PaParaZz1) [inline ding/model/common/head.py:266]: shape (4, 10) -> (4, 5, 2)?

File: ding/model/common/head.py [hunk 11]
```diff
@@ -174,6 +174,108 @@ def forward(self, x: torch.Tensor) -> Dict:
...
+        """
+        value_out = self.V(x)
+        value_out = torch.unsqueeze(value_out, 1)
+        action_out = []
+        for b in self.branches:
+            action_out.append(b(x))
+        action_scores = torch.stack(action_out, 1)
+        action_scores = action_scores - torch.mean(action_scores, 2, keepdim=True)
```
Comment (PaParaZz1) [inline ding/model/common/head.py:274]: add comments for why substracts mean here

File: ding/model/template/q_learning.py [hunk 12]
```diff
@@ -98,6 +98,98 @@ def forward(self, x: torch.Tensor) -> Dict:
...
+            a_layer_num: Optional[int] = None,
+            v_layer_num: Optional[int] = None,
+            encoder_hidden_size_list: SequenceType = [128, 128, 64],
+            head_hidden_size: Optional[int] = None,
+            norm_type: Optional[nn.Module] = None,
+            activation: Optional[nn.Module] = nn.ReLU(),
+    ) -> None:
+        """
+        Overview:
+            Init the BDQ (encoder + head) Model according to input arguments.
```
Comment (PaParaZz1) [inline ding/model/template/q_learning.py:119]: add paper link

File: ding/model/template/q_learning.py [hunk 13]
```diff
@@ -98,6 +98,98 @@ def forward(self, x: torch.Tensor) -> Dict:
...
+        Arguments:
+            - obs_shape (:obj:`Union[int, SequenceType]`): Observation space shape, such as 8 or [4, 84, 84].
+            - num_branches (:obj:`int`): The number of branches, which is equivalent to the action dimension, such as 6.
```
Comment (PaParaZz1) [inline ding/model/template/q_learning.py:122]: you need to specify which env owns 6 action here

File: ding/rl_utils/td.py [hunk 14]
```diff
@@ -453,6 +453,62 @@ def q_nstep_td_error(
     return (td_error_per_sample * weight).mean(), td_error_per_sample
 
 
+def bdq_nstep_td_error(
```
Comment (PaParaZz1) [inline ding/rl_utils/td.py:456]: add unittest for this file

File: dizoo/mujoco/config/halfcheetah_bdq_config.py [hunk 15]
```diff
@@ -0,0 +1,71 @@
+from easydict import EasyDict
+
+halfcheetah_bdq_config = dict(
+    exp_name='halfcheetah_bdq_seed0',
+    env=dict(
+        env_id='HalfCheetah-v3',
+        norm_obs=dict(use_norm=False, ),
```
Comment (PaParaZz1) [inline dizoo/mujoco/config/halfcheetah_bdq_config.py:7]: remove this two lines, they are unused now

File: dizoo/mujoco/config/halfcheetah_bdq_config.py [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+            num_branches=6,
+            action_per_branch=2,  # mean the action shape is 11, 11 discrete actions
+            encoder_hidden_size_list=[256, 256, 128],
+        ),
+        
+        learn=dict(
+            batch_size=512,
+            learning_rate=3e-4,
+            ignore_done=True,
+            target_update_freq=500,
```
Comment (PaParaZz1) [inline dizoo/mujoco/config/halfcheetah_bdq_config.py:31]: why specify these two `target_update_xxx` arguments together

File: dizoo/mujoco/config/halfcheetah_bdq_config.py [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+    # env_manager=dict(type='subprocess'),
+    policy=dict(type='bdq', ),
+)
+halfcheetah_bdq_create_config = EasyDict(halfcheetah_bdq_create_config)
+create_config = halfcheetah_bdq_create_config
+
+if __name__ == "__main__":
+    # or you can enter `ding -m serial_onpolicy -c halfcheetah_onbdq_config.py -s 0`
+    from ding.entry import serial_pipeline
+    serial_pipeline((main_config, create_config), seed=0)
```
Comment (PaParaZz1) [inline dizoo/mujoco/config/halfcheetah_bdq_config.py:71]: specify`max_env_step`

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: ding/model/common/head.py (hunk 11 above)
Comment (Cloud-Pku) [reply ding/model/common/head.py:274]: The original paper explains that implementation (1) is better than implementation (2), (3)

image
image

File: dizoo/mujoco/config/halfcheetah_bdq_config.py (hunk 15 above)
Comment (Cloud-Pku) [reply dizoo/mujoco/config/halfcheetah_bdq_config.py:7]: If the 'norm_reward' parameter is not declared, the following error is displayed.

  File "/DI-engine/dizoo/mujoco/envs/mujoco_env.py", line 154, in create_evaluator_env_cfg
    evaluator_cfg.norm_reward.use_norm = False
AttributeError: 'EasyDict' object has no attribute 'norm_reward'

File: ding/policy/bdq.py (hunk 3 above)
Comment (Cloud-Pku) [reply ding/policy/bdq.py:242]: I've added q_value of every branch

**Review by PaParaZz1 — APPROVED** [review]

**Pull request merged**